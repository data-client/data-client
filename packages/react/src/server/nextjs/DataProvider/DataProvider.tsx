'use client';
import { useMemo, type ComponentProps } from 'react';

import createPersistedStore from './createPersistedStore.js';
import ServerDataComponent from './ServerDataComponent.js';
import type CacheProvider from '../../../components/CacheProvider.js';

export default function DataProvider({
  children,
  ...props
}: ProviderProps): React.ReactElement {
  const [ServerCacheProvider, initPromise] = useMemo(createPersistedStore, []);

  return (
    <>
      <ServerDataComponent initPromise={initPromise} />
      <ServerCacheProvider {...props} initPromise={initPromise}>
        {children}
      </ServerCacheProvider>
    </>
  );
}

type ProviderProps = Omit<
  Partial<ComponentProps<typeof CacheProvider>>,
  'initialState'
> & {
  children: React.ReactNode;
};
