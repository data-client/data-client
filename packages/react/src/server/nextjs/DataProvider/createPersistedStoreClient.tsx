/// <reference types="react/canary" />
import { type State } from '@data-client/core';
import { ComponentProps, use } from 'react';

import CacheProvider from '../../../components/CacheProvider.js';
import { awaitInitialData } from '../../getInitialData.js';

export default function createPersistedStore() {
  const initPromise = awaitInitialData();

  const StoreCacheProvider = ({
    children,
    initPromise,
    ...props
  }: ProviderProps) => {
    const initialState = use(initPromise);
    return (
      <CacheProvider {...props} initialState={initialState}>
        {children}
      </CacheProvider>
    );
  };
  return [StoreCacheProvider, initPromise] as const;
}

type ProviderProps = Omit<
  Partial<ComponentProps<typeof CacheProvider>>,
  'initialState'
> & {
  children: React.ReactNode;
  initPromise: Promise<State<any>>;
};
