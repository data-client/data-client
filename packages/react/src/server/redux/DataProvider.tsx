'use client';
import type { Controller, Manager, State } from '@data-client/core';
import React, { useEffect, useMemo } from 'react';

import ExternalCacheProvider from './ExternalDataProvider.js';
import { prepareStore } from './prepareStore.js';

/** For usage with https://dataclient.io/docs/api/makeRenderDataClient */
export default function ExternalDataProvider({
  children,
  managers,
  initialState,
  Controller,
}: Props) {
  const { selector, store, controller } = useMemo(
    () => prepareStore(initialState, managers, Controller),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [Controller, ...managers],
  );

  // this is not handled in ExternalCacheProvider as it doesn't
  // own its managers. Since we are owning them here, we should ensure it happens
  useEffect(() => {
    for (let i = 0; i < managers.length; ++i) {
      managers[i].init?.(selector(store.getState()));
    }
    return () => {
      for (let i = 0; i < managers.length; ++i) {
        managers[i].cleanup();
      }
    };
    // we're ignoring state here, because it shouldn't trigger inits
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [...managers]);

  return (
    <ExternalCacheProvider
      store={store}
      selector={selector}
      controller={controller}
    >
      {children}
    </ExternalCacheProvider>
  );
}

interface Props {
  children: React.ReactNode;
  managers: Manager[];
  initialState: State<unknown>;
  Controller: typeof Controller;
}
