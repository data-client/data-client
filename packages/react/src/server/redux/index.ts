export { applyManager, initialState, createReducer } from '@data-client/core';
// export { prepareStore } from './prepareStore.js';
// export { default as CacheProvider } from './CacheProvider.js';
export { default as ExternalCacheProvider } from './ExternalCacheProvider.js';
export { default as mapMiddleware } from './mapMiddleware.js';
export { default as PromiseifyMiddleware } from './PromiseifyMiddleware.js';
