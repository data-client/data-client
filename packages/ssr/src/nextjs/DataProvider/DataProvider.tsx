'use client';
import { type DataProvider as DataProviderClient } from '@data-client/react';
import { useMemo, type ComponentProps } from 'react';

import createPersistedStore from './createPersistedStore.js';
import ServerDataComponent from './ServerDataComponent.js';

export default function DataProvider({
  children,
  ...props
}: ProviderProps): React.ReactElement {
  const [ServerDataProvider, initPromise] = useMemo(createPersistedStore, []);

  return (
    <>
      <ServerDataComponent initPromise={initPromise} />
      <ServerDataProvider {...props} initPromise={initPromise}>
        {children}
      </ServerDataProvider>
    </>
  );
}

type ProviderProps = Omit<
  Partial<ComponentProps<typeof DataProviderClient>>,
  'initialState'
> & {
  children: React.ReactNode;
};
