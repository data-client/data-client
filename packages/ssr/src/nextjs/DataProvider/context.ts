'use client';
import { type State, __INTERNAL__ } from '@data-client/react';
import { createContext } from 'react';

export const readyContext = createContext(
  (): State<unknown> => __INTERNAL__.initialState,
);
