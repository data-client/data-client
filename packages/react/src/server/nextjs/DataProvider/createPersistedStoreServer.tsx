import {
  Controller,
  Manager,
  NetworkManager,
  State,
  createReducer,
  initialState,
  applyManager,
} from '@data-client/core';
import type { ComponentProps } from 'react';

import type DataProvider from '../../../components/DataProvider.js';
import {
  ExternalDataProvider,
  PromiseifyMiddleware,
} from '../../redux/index.js';
import { createStore, applyMiddleware } from '../../redux/redux.js';

export default function createPersistedStore(managers?: Manager[]) {
  const controller = new Controller();
  managers = managers ?? [new NetworkManager()];
  const nm: NetworkManager = managers.find(
    m => m instanceof NetworkManager,
  ) as any;
  if (nm === undefined)
    throw new Error('managers must include a NetworkManager');
  const reducer = createReducer(controller);
  const enhancer = applyMiddleware(
    // redux 5's types are wrong and do not allow any return typing from next, which is incorrect.
    // `next: (action: unknown) => unknown`: allows any action, but disallows all return types.
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    ...applyManager(managers, controller),
    PromiseifyMiddleware,
  );
  const store = createStore(reducer, initialState as any, enhancer);
  managers.forEach(manager => manager.init?.(store.getState()));

  const selector = (state: any) => state;

  const getState = () => selector(store.getState());

  const initPromise: Promise<State<any>> = (async () => {
    let firstRender = true;
    // eslint-disable-next-line no-constant-condition
    while (true) {
      const inFlightFetches = nm.allSettled();
      if (inFlightFetches) {
        firstRender = false;
        await inFlightFetches;
        continue;
      }
      if (firstRender) {
        firstRender = false;
        // TODO: instead of waiting 10ms - see if we can wait until next part of react is streamed and race with nm getting new fetches
        await new Promise(resolve => setTimeout(resolve, 10));
        continue;
      }
      break;
    }
    return getState();
  })();

  const StoreDataProvider = ({ children }: ProviderProps) => {
    return (
      <ExternalDataProvider
        store={store}
        selector={selector}
        controller={controller}
      >
        {children}
      </ExternalDataProvider>
    );
  };

  return [StoreDataProvider, initPromise] as const;
}

type ProviderProps = Omit<
  Partial<ComponentProps<typeof DataProvider>>,
  'initialState'
> & {
  children: React.ReactNode;
  initPromise: Promise<State<any>>;
};
