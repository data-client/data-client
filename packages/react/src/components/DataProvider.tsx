'use client';
import {
  initialState as defaultState,
  Controller as DataController,
  applyManager,
} from '@data-client/core';
import type { State, Manager } from '@data-client/core';
import React, { useCallback, useMemo, useRef } from 'react';
import type { JSX } from 'react';

import DataStore from './DataStore.js';
import type { DevToolsPosition } from './DevToolsButton.js';
import { getDefaultManagers } from './getDefaultManagers.js';
import { SSR } from './LegacyReact.js';
import { renderDevButton } from './renderDevButton.js';
import { ControllerContext } from '../context.js';
import { DevToolsManager } from '../managers/index.js';

export interface ProviderProps {
  children: React.ReactNode;
  managers?: Manager[];
  initialState?: State<unknown>;
  Controller?: typeof DataController;
  devButton?: DevToolsPosition | null | undefined;
}

interface Props {
  children: React.ReactNode;
  managers?: Manager[];
  initialState?: State<unknown>;
  Controller?: typeof DataController;
  devButton?: DevToolsPosition | null | undefined;
}

/**
 * Manages state, providing all context needed to use the hooks.
 * @see https://dataclient.io/docs/api/DataProvider
 */
export default function DataProvider({
  children,
  managers,
  initialState = defaultState as State<unknown>,
  Controller = DataController,
  devButton = 'bottom-right',
}: Props): JSX.Element {
  /* istanbul ignore else */
  if (process.env.NODE_ENV !== 'production' && SSR) {
    console.warn(
      `DataProvider from @data-client/react does not update while doing SSR.
See https://dataclient.io/docs/guides/ssr.`,
    );
  }
  // contents of this component expected to be relatively stable
  const controllerRef: React.MutableRefObject<DataController> =
    useRef<any>(undefined);
  if (!controllerRef.current) controllerRef.current = new Controller();
  //TODO: bind all methods so destructuring works

  const managersRef: React.MutableRefObject<Manager[]> = useRef<any>(managers);
  if (!managersRef.current) managersRef.current = getDefaultManagers();

  // run in a useEffect in DataStore
  const mgrEffect = useCallback(() => {
    managersRef.current.forEach(manager => {
      manager.init?.(initialState);
    });
    return () => {
      managersRef.current.forEach(manager => {
        manager.cleanup();
      });
    };
    // we don't support initialState changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, managersRef.current);

  // Makes manager middleware compatible with redux-style middleware (by a wrapper enhancement to provide controller API)
  const middlewares = useMemo(
    () => applyManager(managersRef.current, controllerRef.current),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    managersRef.current,
  );

  // only include if they have devtools integrated
  const hasDevManager = !!managersRef.current.find(
    manager => manager instanceof DevToolsManager,
  );
  return (
    <ControllerContext.Provider value={controllerRef.current}>
      <DataStore
        mgrEffect={mgrEffect}
        middlewares={middlewares}
        initialState={initialState}
        controller={controllerRef.current}
      >
        {children}
      </DataStore>
      {renderDevButton(devButton, hasDevManager)}
    </ControllerContext.Provider>
  );
}
