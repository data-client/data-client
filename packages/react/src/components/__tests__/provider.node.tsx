// eslint-env jest
import React, { version } from 'react';
import { renderToString } from 'react-dom/server';

import DataProvider from '../DataProvider';

describe('<BackupBoundary />', () => {
  let warnspy: jest.SpyInstance;
  beforeEach(() => {
    warnspy = jest.spyOn(global.console, 'warn').mockImplementation(() => {});
  });
  afterEach(() => {
    warnspy.mockRestore();
  });

  it('should warn users about SSR with DataProvider', async () => {
    const tree = (
      <DataProvider>
        <div>hi</div>
      </DataProvider>
    );
    const LegacyReact = version.startsWith('16') || version.startsWith('17');

    // let our lazy component load
    renderToString(tree);
    await new Promise(resolve => setTimeout(resolve, 0));

    const msg = renderToString(tree);

    expect(warnspy.mock.lastCall).toMatchSnapshot();

    if (!LegacyReact) {
      expect(msg).toMatchInlineSnapshot(
        `"<!--$--><div>hi</div><!--/$--><!--$--><!--/$-->"`,
      );
    } else {
      expect(msg).toMatchInlineSnapshot(`"<div>hi</div>"`);
    }
  });
});
