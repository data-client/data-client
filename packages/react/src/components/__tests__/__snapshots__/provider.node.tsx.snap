// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BackupBoundary /> should warn users about SSR with DataProvider 1`] = `
[
  "DataProvider from @data-client/react does not update while doing SSR.
See https://dataclient.io/docs/guides/ssr.",
]
`;
